File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
#include <unordered_map>
#include <algorithm>
#include <iostream>
#include <string>


class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> map;
        
        int max_so_far = INT_MIN;
        int count = 0;
        int begin = 0;
        int end = 0;
        int row = 0;
        const int S_LEN = s.length();
    
        
        while(end < S_LEN) {
            if (map.find(s[end]) == map.end() || map.find(s[end])->second == 0) {
                map[s[end]]++;
                end++;
                count++;
            }
            
            else {
                max_so_far = std::max(count, max_so_far);
                map[s[begin]] = 0;
                begin++;
                count--;

            }
        }
        
        return count > max_so_far ? count : max_so_far;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        l = 0
        res = 0
        count = collections.defaultdict(int)
        
        for r in range(len(s)):
            count[s[r]]+=1
            
            while count[s[r]] > 1:
                count[s[l]]-=1
                l+=1
            
            res = max(res, r-l+1)
                
        return res